From 13cfa364d81395e337a0479739846c015d29115c Mon Sep 17 00:00:00 2001
From: Zongdong Jiao <zongdong.jiao@amlogic.com>
Date: Tue, 6 Jan 2015 10:33:58 +0800
Subject: [PATCH 5510/5965] PD#101764: hdmi: disable phy while dynamic switch
 to DVI mode

For some DVI devices, they can not recognize signal from HDMI mode to
DVI mode, and driver must be disable PHY first and enable PHY again.

Change-Id: Ib933198ba78b15f52fb9dd8978b833ba5596b8d3
---
 arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c  | 3 +++
 arch/arm/mach-meson8b/hdmi_tx_hw/hdmi_tx_hw.c | 3 +++
 2 files changed, 6 insertions(+)

diff --git a/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c b/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c
index 5a4c1bbbe633..2ad3010ee3b9 100755
--- a/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c
+++ b/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c
@@ -3513,8 +3513,11 @@ static int hdmitx_cntl_config(hdmitx_dev_t* hdmitx_device, unsigned cmd, unsigne
             hdmi_set_reg_bits(TX_TMDS_MODE, 0x3, 6, 2);
         }
         if(argv == DVI_MODE) {
+            hdmi_phy_suspend();
             hdmi_set_reg_bits(TX_VIDEO_DTV_OPTION_L, 0x0, 6, 2);
             hdmi_set_reg_bits(TX_TMDS_MODE, 0x2, 6, 2);
+            msleep(300);
+            hdmi_phy_wakeup(hdmitx_device);
         }
         break;
     case CONF_SYSTEM_ST:
diff --git a/arch/arm/mach-meson8b/hdmi_tx_hw/hdmi_tx_hw.c b/arch/arm/mach-meson8b/hdmi_tx_hw/hdmi_tx_hw.c
index 4ecc4f9c4caa..4e2ef086bbe7 100755
--- a/arch/arm/mach-meson8b/hdmi_tx_hw/hdmi_tx_hw.c
+++ b/arch/arm/mach-meson8b/hdmi_tx_hw/hdmi_tx_hw.c
@@ -3290,8 +3290,11 @@ static int hdmitx_cntl_config(hdmitx_dev_t* hdmitx_device, unsigned cmd, unsigne
             hdmi_set_reg_bits(TX_TMDS_MODE, 0x3, 6, 2);
         }
         if(argv == DVI_MODE) {
+            hdmi_phy_suspend();
             hdmi_set_reg_bits(TX_VIDEO_DTV_OPTION_L, 0x0, 6, 2);
             hdmi_set_reg_bits(TX_TMDS_MODE, 0x2, 6, 2);
+            msleep(300);
+            hdmi_phy_wakeup(hdmitx_device);
         }
         break;
     case CONF_SYSTEM_ST:
-- 
2.19.0

